src/event-log.c

```diff
@@ -0,0 +1,398 @@
+			formatted = event_log_format_fields_readable(G_LOG_LEVEL_MESSAGE, fields, n_fields, FALSE);
+			break;
+		case R_EVENT_LOGFMT_JSON:
+#if ENABLE_JSON
+			formatted = event_log_format_fields_json(G_LOG_LEVEL_MESSAGE, fields, n_fields, FALSE);
+#else
+			g_error("Compiled without JSON support");
+#endif
+			break;
+		case R_EVENT_LOGFMT_JSON_PRETTY:
+#if ENABLE_JSON
+			formatted = event_log_format_fields_json(G_LOG_LEVEL_MESSAGE, fields, n_fields, TRUE);
+#else
+			g_error("Compiled without JSON support");
+#endif
+			break;
+		default:
+			g_error("Unknown log format");
+	}
+	output = g_strdup_printf("%s\n", formatted);
+
+	/* Once we know how much to write, we can use this information for trimming */
+	rotate_logfiles(logger, strlen(output));
+
+	if (!g_output_stream_write_all(G_OUTPUT_STREAM(logger->logstream), output, strlen(output), &written, NULL, &ierror)) {
+		g_warning("Failed to write log file: %s", ierror->message);
+		return;
+	}
+
+	logger->filesize += written;
+}
+
+GLogWriterOutput r_event_log_writer(GLogLevelFlags log_level, const GLogField *fields, gsize n_fields, gpointer user_data)
+{
+	const gchar *log_domain = NULL;
+	const gchar *event_type = NULL;
+
+	/* Always log to default location, too */
+	g_log_writer_default(log_level, fields, n_fields, user_data);
+
+	/* get log domain */
+	for (gsize i = 0; i < n_fields; i++) {
+		if (g_strcmp0(fields[i].key, "GLIB_DOMAIN") == 0) {
+			log_domain = fields[i].value;
+			break;
+		}
+	}
+
+	/* We are interested in "rauc-event" domains only */
+	if (!log_domain || g_strcmp0(log_domain, R_EVENT_LOG_DOMAIN) != 0) {
+		return G_LOG_WRITER_HANDLED;
+	}
+
+	/* get event type */
+	for (gsize i = 0; i < n_fields; i++) {
+		if (g_strcmp0(fields[i].key, "RAUC_EVENT_TYPE") == 0) {
+			event_type = fields[i].value;
+			break;
+		}
+	}
+
+	/* iterate over registered event loggers */
+	for (GList *l = r_context()->config->loggers; l != NULL; l = l->next) {
+		REventLogger* logger = l->data;
+
+		/* Filter out by event type */
+		if ((g_strcmp0(logger->events[0], "all") != 0) &&
+		    !g_strv_contains((const gchar * const*)logger->events, event_type)) {
+			continue;
+		}
+
+		logger->writer(logger, fields, n_fields);
+	}
+
+	return G_LOG_WRITER_HANDLED;
+}
+
+void r_event_log_setup_logger(REventLogger *logger)
+{
+	g_autoptr(GError) ierror = NULL;
+
+	g_return_if_fail(logger);
+
+	if (logger->configured) {
+		g_message("Logger %s already configured", logger->name);
+		return;
+	}
+
+	g_info("Setting up logger %s for %s ..", logger->name, logger->filename);
+
+	logger->writer = &event_log_writer_file;
+
+	/* query file information or create new ones */
+	if (!g_file_test(logger->filename, G_FILE_TEST_EXISTS)) {
+		logger->filesize = 0;
+	} else {
+		g_autoptr(GFile) logfile = NULL;
+		g_autoptr(GFileInfo) logfile_info = NULL;
+
+		logfile = g_file_new_for_path(logger->filename);
+		logfile_info = g_file_query_info(logfile, G_FILE_ATTRIBUTE_STANDARD_SIZE, G_FILE_QUERY_INFO_NONE, NULL, &ierror);
```

Comment:
Couldn't this be done using `g_seekable_tell()` after opening the file?

Reply:
Already opening the logfile in the setup probably makes sense anyway. Since `GFileOutputStream` implements `GSeekable` this should work, yes.